nifi-nar-bundles/nifi-snowflake-bundle/nifi-snowflake-services/src/main/java/org/apache/nifi/snowflake/service/SnowflakeComputingConnectionPool.java

```diff
@@ -0,0 +1,314 @@
+        final String validationQuery = context.getProperty(VALIDATION_QUERY).evaluateAttributeExpressions().getValue();
+        final Long maxWaitMillis = extractMillisWithInfinite(context.getProperty(MAX_WAIT_TIME).evaluateAttributeExpressions());
+        final Integer minIdle = context.getProperty(MIN_IDLE).evaluateAttributeExpressions().asInteger();
+        final Integer maxIdle = context.getProperty(MAX_IDLE).evaluateAttributeExpressions().asInteger();
+        final Long maxConnLifetimeMillis = extractMillisWithInfinite(context.getProperty(MAX_CONN_LIFETIME).evaluateAttributeExpressions());
+        final Long timeBetweenEvictionRunsMillis = extractMillisWithInfinite(context.getProperty(EVICTION_RUN_PERIOD).evaluateAttributeExpressions());
+        final Long minEvictableIdleTimeMillis = extractMillisWithInfinite(context.getProperty(MIN_EVICTABLE_IDLE_TIME).evaluateAttributeExpressions());
+        final Long softMinEvictableIdleTimeMillis = extractMillisWithInfinite(context.getProperty(SOFT_MIN_EVICTABLE_IDLE_TIME).evaluateAttributeExpressions());
+
+        dataSource = new BasicDataSource();
+
+        dataSource.setDriver(getDriver(SnowflakeDriver.class.getName(), connectionString));
+
+        dataSource.setUrl(connectionString);
+        dataSource.setUsername(user);
+        dataSource.setPassword(password);
+
+        dataSource.setMaxWaitMillis(maxWaitMillis);
+        dataSource.setMaxTotal(maxTotal);
+        dataSource.setMinIdle(minIdle);
+        dataSource.setMaxIdle(maxIdle);
+        dataSource.setMaxConnLifetimeMillis(maxConnLifetimeMillis);
+        dataSource.setTimeBetweenEvictionRunsMillis(timeBetweenEvictionRunsMillis);
+        dataSource.setMinEvictableIdleTimeMillis(minEvictableIdleTimeMillis);
+        dataSource.setSoftMinEvictableIdleTimeMillis(softMinEvictableIdleTimeMillis);
+
+        if (validationQuery != null && !validationQuery.isEmpty()) {
+            dataSource.setValidationQuery(validationQuery);
+            dataSource.setTestOnBorrow(true);
+        }
+
+        final List<PropertyDescriptor> dynamicProperties = context.getProperties()
+            .keySet()
+            .stream()
+            .filter(PropertyDescriptor::isDynamic)
+            .collect(Collectors.toList());
+
+        dynamicProperties.forEach((descriptor) -> {
+            final PropertyValue propertyValue = context.getProperty(descriptor);
+            if (descriptor.isSensitive()) {
+                final String propertyName = StringUtils.substringAfter(descriptor.getName(), SENSITIVE_PROPERTY_PREFIX);
+                dataSource.addConnectionProperty(propertyName, propertyValue.getValue());
+            } else {
+                dataSource.addConnectionProperty(descriptor.getName(), propertyValue.evaluateAttributeExpressions().getValue());
+            }
+        });
+    }
+
+    protected Driver getDriver(final String driverName, final String url) {
```

Comment:
Wouldn't this make the new module expose the ```DBCPConnectionPool``` service a second time?

Reply:
Are you referring to exposing the `DBCPConnectionPool` to the list of available Controller Services through the `nifi-snowflake-services-nar`? I did not observe that behavior in runtime testing.

Reply:
For me this change results in an "Incompatible Controller Service Configured" error on the processor.
Were you able to run a successful query with this setup? Is there something else that needs to be changed?

Reply:
I was able to configure and enable the `SnowflakeComputingConnectionPool` service and attempting a query using `ExecuteSQL` without any incompatible errors.

As an even better solution, instead of adding `cloneAncestorResources = true`, overriding the `getDriver()` method to return a new instance of the `SnowflakeDriver` class also appears to work, since it avoids the `Class.forName()` lookup in `DBCPConnectionPool`.  The method in the SnowflakeComputing class would read as follows:

```suggestion
    @Override
    protected Driver getDriver(final String driverName, final String url) {
        return new SnowflakeDriver();
    }
```

Reply:
I'm not sure it's a good idea to bypass the ```DriverManager``` though.
I think we should keep the logic that returns the (same) registered driver for each call (and also checks that the URL is compatible).

That's really all the ```DBCPConnectionPool.getDriver()``` itself does with some exception handling.

Reply:
The Commons DBCP BasicDataSource does not need `DriverManager` when it has a `Driver` instance, and the `SnowflakeDriver` instance could be declared as a static variable.  The implied URL checking is useful, so that is an option.  Either approach to providing a simple method override would be good, as opposed to duplication.